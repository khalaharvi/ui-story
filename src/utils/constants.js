export const PRIMARY_CATEGORY_KEY = 'categories.lvl0';
export const SECONDARY_CATEGORY_KEY = 'categories.lvl1';
export const TERTIARY_CATEGORY_KEY = 'categories.lvl2';
