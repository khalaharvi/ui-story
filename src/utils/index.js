export { default as createClassNames } from './createClassNames';
export { default as getDisplayName } from './getDisplayName';
export { default as capitalizeFirstLetter } from './capitalizeFirstLetter';
export { default as slugify } from './slugify';
