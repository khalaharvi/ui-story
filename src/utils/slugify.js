export default (text) => {
  // Use hash map for special characters
  const specialChars = {"à":'a',"ä":'a',"á":'a',"â":'a',"æ":'a',"å":'a',"ë":'e',"è":'e',"é":'e', "ê":'e',"î":'i',"ï":'i',"ì":'i',"í":'i',"ò":'o',"ó":'o',"ö":'o',"ô":'o',"ø":'o',"ù":'o',"ú":'u',"ü":'u',"û":'u',"ñ":'n',"ç":'c',"ß":'s',"ÿ":'y',"œ":'o',"ŕ":'r',"ś":'s',"ń":'n',"ṕ":'p',"ẃ":'w',"ǵ":'g',"ǹ":'n',"ḿ":'m',"ǘ":'u',"ẍ":'x',"ź":'z',"ḧ":'h',"·":'-',"/":'-',"_":'-',",":'-',":":'-',";":'-'}; // eslint-disable-line

  return text.toString().toLowerCase()
    .replace(/\s+/g, '-') // Replace spaces with -
    .replace(/./g, target => specialChars[target] || target) // Replace special characters using the hash map
    .replace(/&/g, '-and-') // Replace & with 'and'
    .replace(/[^\w-]+/g, '') // Remove all non-word chars
    .replace(/--+/g, '-') // Replace multiple - with single -
    .replace(/^-+/, '') // Trim - from start of text
    .replace(/-+$/, ''); // Trim - from end of text
};
