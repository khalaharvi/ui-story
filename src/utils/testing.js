import { object } from 'prop-types';
import { channel, createBroadcast } from 'emotion-theming';
import { shallow, mount } from 'enzyme';
import theme from '@dering-hall/dh-theme';

const getOptions = (options = {}) => {
  const broadcast = createBroadcast(theme);

  return {
    context: {
      [channel]: broadcast
    },
    childContextTypes: {
      [channel]: object
    },
    ...options
  };
};

export const mountWithTheme = (
  Component,
  options = {}
) => mount(Component, { ...getOptions(options) });

export const shallowWithTheme = (
  Component,
  options = {}
) => shallow(Component, { ...getOptions(options) });

export default {
  mountWithTheme,
  shallowWithTheme
};
