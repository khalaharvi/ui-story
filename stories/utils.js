/* eslint-disable import/prefer-default-export */
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { InstantSearch } from 'react-instantsearch/dom';
import { connectStateResults } from 'react-instantsearch/connectors';
import { withMedia, MediaQueryProvider } from 'react-media-query-hoc';
import { withTests as withTestsAddon } from '@storybook/addon-jest';
import { ThemeProvider } from 'emotion-theming';
import dhTheme from '@dering-hall/dh-theme';
import testsResults from '../.jest-test-results.json';

const Wrap = ({ index, children }) => (
  <InstantSearch
    appId={process.env.ALGOLIA_ID}
    apiKey={process.env.ALGOLIA_API_KEY}
    indexName={index}
  >
    {children}
  </InstantSearch>
);

Wrap.propTypes = {
  index: PropTypes.string,
  children: PropTypes.node
};

Wrap.defaultProps = {
  index: process.env.ALGOLIA_PRODUCT_INDEX
};

const WrapThemeProvider = ({ children, theme }) => (
  <ThemeProvider theme={theme}>
    <Fragment>
      {children}
    </Fragment>
  </ThemeProvider>
);

WrapThemeProvider.propTypes = {
  theme: PropTypes.object // eslint-disable-line
};

WrapThemeProvider.defaultProps = {
  theme: dhTheme
};

const SearchStateDebug = connectStateResults(({ searchState }) => (
  <div style={{ margin: '1rem 0', fontFamily: 'monospace' }}>
    <pre
      style={{
        background: '#f6f8fa',
        fontSize: '.65rem',
        padding: '.5rem'
      }}
    >
      <strong>searchState</strong> = {JSON.stringify(searchState, null, '\t')}
    </pre>
  </div>
));

const customQueries = {
  small: 'screen and (max-width: 730px)',
  medium: 'screen and (max-width: 970px)',
  large: 'screen and (max-width: 1240px)',
  xlarge: 'screen and (min-width: 1240px)'
};

const MediaQueriesWrapper = ({ children }) => (
  <MediaQueryProvider queries={customQueries}>
    {children}
  </MediaQueryProvider>
);

const OnSmallOnly = withMedia(({ media, children }) => {
  if (!media.small) {
    return null;
  }

  return (
    <Fragment>
      {children}
    </Fragment>
  );
});

const withTests = withTestsAddon({
  results: testsResults
});

export {
  SearchStateDebug,
  MediaQueriesWrapper,
  OnSmallOnly,
  Wrap,
  WrapThemeProvider,
  withTests
};
/* eslint-enable */
