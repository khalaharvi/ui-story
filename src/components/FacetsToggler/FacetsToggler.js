import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { createClassNames } from 'utils';
import Button from 'components/Button';

const cx = createClassNames('facets-toggler');
export const defaultToggleClassName = createClassNames('facets')('visible');

class FacetsToggler extends Component {
  static propTypes = {
    toggleClassName: PropTypes.string
  };

  state = {
    visible: false
  };

  componentDidMount() {
    this.body = document.body;
  }

  componentWillUnmount() {
    this.body.classList.remove(...this.getVisibleClassnames());
  }

  onClick = () => {
    const { visible } = this.state;

    this.setState({
      visible: !visible
    }, () => {
      if (visible) {
        this.body.classList.remove(...this.getVisibleClassnames());
      } else {
        this.body.classList.add(...this.getVisibleClassnames());
      }
    });
  };

  getVisibleClassnames() {
    const classNames = [defaultToggleClassName];
    const { toggleClassName } = this.props;

    if (toggleClassName) {
      classNames.push(toggleClassName);
    }

    return classNames;
  }

  render() {
    const { visible } = this.state;
    const { className } = this.props;

    return (
      <div className={classnames(cx(''), className)}>
        <Button
          className={cx('toggle')}
          onClick={this.onClick}
          label={`${visible ? 'Hide' : 'Refine'} Search`}
          type="secondary"
        />
      </div>
    );
  }
}

export default FacetsToggler;
