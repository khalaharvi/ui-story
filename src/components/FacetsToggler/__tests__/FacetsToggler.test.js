import React from 'react';
import { shallow } from 'enzyme';
import { mountWithTheme } from 'utils/testing';

import FacetsToggler, { defaultToggleClassName } from '../FacetsToggler';

// Create a div in the document to mount into
const div = global.document.createElement('div');
global.document.body.appendChild(div);

describe('<FacetsToggler />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<FacetsToggler />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with a custom className', () => {
    const wrapper = shallow(<FacetsToggler className="custom-classname" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should update its state and <body> classnames when clicked', () => {
    const wrapper = mountWithTheme(<FacetsToggler />, { attachTo: div });

    const button = wrapper.find('button');

    button.simulate('click');

    expect(document.body.classList.contains(defaultToggleClassName)).toBe(true);
    expect(wrapper.state()).toEqual({
      visible: true
    });

    button.simulate('click');

    expect(document.body.classList.contains(defaultToggleClassName)).toBe(false);
    expect(wrapper.state()).toEqual({
      visible: false
    });

    wrapper.detach();
    wrapper.unmount();
  });

  it('should remove <body> toggle classname on componentWillUnmount', () => {
    const wrapper = mountWithTheme(<FacetsToggler />, { attachTo: div });

    const button = wrapper.find('button');

    button.simulate('click');

    expect(document.body.classList.contains(defaultToggleClassName)).toBe(true);
    expect(wrapper.state()).toEqual({
      visible: true
    });

    wrapper.detach();
    wrapper.unmount();

    expect(document.body.classList.contains(defaultToggleClassName)).toBe(false);
  });

  it('should update <body> classnames with custom toggle className when clicked', () => {
    const wrapper = mountWithTheme(
      <FacetsToggler toggleClassName="custom-toggle-classname" />,
      { attachTo: div }
    );

    const button = wrapper.find('button');

    button.simulate('click');

    expect(document.body.classList.contains(defaultToggleClassName)).toBe(true);
    expect(document.body.classList.contains('custom-toggle-classname')).toBe(true);

    button.simulate('click');

    expect(document.body.classList.contains(defaultToggleClassName)).toBe(false);
    expect(document.body.classList.contains('custom-toggle-classname')).toBe(false);

    wrapper.detach();
    wrapper.unmount();
  });

  it('should remove <body> custom toggle classname on componentWillUnmount', () => {
    const wrapper = mountWithTheme(
      <FacetsToggler toggleClassName="custom-toggle-classname" />,
      { attachTo: div }
    );

    const button = wrapper.find('button');

    button.simulate('click');

    expect(document.body.classList.contains(defaultToggleClassName)).toBe(true);
    expect(document.body.classList.contains('custom-toggle-classname')).toBe(true);

    wrapper.detach();
    wrapper.unmount();

    expect(document.body.classList.contains(defaultToggleClassName)).toBe(false);
    expect(document.body.classList.contains('custom-toggle-classname')).toBe(false);
  });
});
