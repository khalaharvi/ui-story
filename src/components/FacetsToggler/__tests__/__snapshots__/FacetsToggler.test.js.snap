// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FacetsToggler /> renders correctly 1`] = `
<div
  className="dh-facets-toggler"
>
  <ButtonComponent
    className="dh-facets-toggler-toggle"
    disabled={false}
    hidden={false}
    iconAfter={false}
    inverted={false}
    label="Refine Search"
    loading={false}
    onClick={[Function]}
    primary={false}
    secondary={false}
    type="secondary"
    white={false}
  />
</div>
`;

exports[`<FacetsToggler /> renders correctly with a custom className 1`] = `
<div
  className="dh-facets-toggler custom-classname"
>
  <ButtonComponent
    className="dh-facets-toggler-toggle"
    disabled={false}
    hidden={false}
    iconAfter={false}
    inverted={false}
    label="Refine Search"
    loading={false}
    onClick={[Function]}
    primary={false}
    secondary={false}
    type="secondary"
    white={false}
  />
</div>
`;
